mlir/tools/miopen-gen/miopen-gen.cpp

```diff
@@ -803,60 +822,342 @@ static func::FuncOp getMemsetFunc(ModuleOp module, mlir::Type elemType) {
+}
 
-  assert(genConfig.operation.hasValue());
-  std::string funcName =
-      miopen::getNameForConvOpType(genConfig.operation.getValue()).str();
+static void
+createCPUConvWithMLIR(ModuleOp module, func::FuncOp &func,
+                      const miopen::Conv2dGenerator::Config &genConfig) {
+  OpBuilder b(module.getContext());
 
-  funcName += "_cpu";
-  func::FuncOp func = module.lookupSymbol<func::FuncOp>(funcName);
-  if (func) // already exists
-    return func;
+  Block *block = func.addEntryBlock();
+  b.setInsertionPoint(block, block->begin());
 
-  OpBuilder b(module.getContext());
   auto loc = b.getUnknownLoc();
+  auto zeroI16Op = b.create<arith::ConstantIntOp>(loc, 0, b.getIntegerType(16));
+  auto zeroI32Op = b.create<arith::ConstantIntOp>(loc, 0, b.getIntegerType(32));
 
-  mlir::Type elemType = b.getF32Type();
-  mlir::Type outputElemType = b.getF32Type();
-  if (genConfig.dataTypeStr == "i8") {
-    elemType = b.getI8Type();
-    outputElemType = b.getIntegerType(32);
-    assert(genConfig.operation.getValue() == miopen::ConvOpType::Fwd);
+  // Initialize the result tensor
+  mlir::BlockArgument resultTensor;
+  switch (genConfig.operation.getValue()) {
+  case miopen::ConvOpType::Fwd:
+    resultTensor = block->getArgument(2);
+    break;
+  case miopen::ConvOpType::BwdData:
+    resultTensor = block->getArgument(1);
+    break;
+  case miopen::ConvOpType::BwdWeight:
+    resultTensor = block->getArgument(0);
+    break;
   }
+  auto resultType = resultTensor.getType().template dyn_cast<MemRefType>();
+  auto elemType = resultType.getElementType();
+  MemRefType mr5DUnkType = MemRefType::get({-1, -1, -1, -1, -1}, elemType);
+  auto memrefCastOp = b.create<memref::CastOp>(loc, mr5DUnkType, resultTensor);
+  b.create<func::CallOp>(
+      loc, getMemsetFunc(module, elemType),
+      ValueRange{memrefCastOp, zeroI16Op, zeroI16Op, zeroI32Op});
 
-  auto filterDimension = genConfig.filterDimension;
-  auto inputDimension = genConfig.inputDimension;
-  auto outputDimension = genConfig.outputDimension;
-
-  auto filterType = MemRefType::get(filterDimension, elemType);
-  auto inputType = MemRefType::get(inputDimension, elemType);
-  auto outputType = MemRefType::get(outputDimension, outputElemType);
+  AffineExpr heightExpr, widthExpr;
+  AffineMap heightMap, widthMap;
 
-  // Create conv2d_host function
-  miopen::Conv2dGenerator conv2dGenerator(genConfig);
-  bool hasWorkspace = conv2dGenerator.hasWorkspace(b);
-  mlir::Type workspaceArgType;
-  if (hasWorkspace) {
-    workspaceArgType = MemRefType::get(
-        ArrayRef<int64_t>(filterDimension.begin(), filterDimension.end()),
-        b.getF32Type());
+  switch (genConfig.operation.getValue()) {
+  case miopen::ConvOpType::Fwd:
+  case miopen::ConvOpType::BwdWeight:
+    heightExpr = b.getAffineDimExpr(0) * genConfig.strideHeight +
+                 b.getAffineDimExpr(1) * genConfig.dilationHeight -
+                 genConfig.paddingHeightLeft;
```

Comment:
What about right padding? More generally, can we get a comment explaining what these maps represent?

Reply:
Right padding is not needed for index computation here. It's used to get the dimensions of the output tensor though.